app/controllers/api/v2/registration_controller.rb

```diff
@@ -0,0 +1,64 @@
+module Api
+  module V2
+    class RegistrationController < V2::BaseController
+      include Api::Version2
+      include Foreman::Controller::Registration
+      include Foreman::Controller::Parameters::Host
+
+      rescue_from ActiveRecord::RecordNotFound do |error|
+        logger.info "#{error.message} (#{error.class})"
+        not_found(error.message)
+      end
+
+      api :GET, '/register', N_('Render Global registration template')
+      param :organization_id, :number, desc: N_("ID of the Organization to register the host in.")
+      param :location_id, :number, desc: N_("ID of the Location to register the host in.")
+      param :hostgroup_id, :number, desc: N_("ID of the Host group to register the host in.")
+      param :operatingsystem_id, :number, desc: N_("ID of the Operating System to register the host in.")
+      def global
+        find_global_registration
+
+        unless @provisioning_template
+          not_found _('Global Registration Template with name %s defined via default_global_registration_item Setting not found, please configure the existing template name first') % Setting[:default_global_registration_item]
+          return
+        end
+
+        render plain: @provisioning_template.render(variables: @global_registration_vars).html_safe
+      end
+
+      api :POST, "/register", N_("Find or create a host and render the Host registration template")
+      param_group :host, Api::V2::HostsController
+      def host
+        begin
+          ActiveRecord::Base.transaction do
```

Comment:
This does not work, `@host.registration_template` does not raise the exception. We need to make sure to rollback if `@template` is nil

```ruby
ActiveRecord::Base.transaction do
  find_host
  prepare_host
  @template = @host.registration_template
  raise ActiveRecord::Rollback if @template.nil?
end
```

this worked for me